### PR TITLE
panos_dag_tags: create tags for dynamic address objects

### PR DESCRIPTION
**Pull request by ivanbojer**
##### SUMMARY
Security policies allow you to enforce rules and take action, and can be as
general or specific as needed. The policy rules are compared against the
incoming traffic in sequence, and because the first rule that matches the
traffic is applied, the more specific rules must precede the more general ones.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
panos_dag_tags

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Feb  6 2017, 11:22:10) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION
N/A

### CODE REVIEW

**Comment by jtschichold** [comment]
:shipit:

File: lib/ansible/modules/network/panos/panos_dag_tags.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    if exc:
+        return False, exc
+    else:
+        return True, exc
+
+
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        password=dict(required=True),
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_dag_tags.py]: please add `,no_log=True`

File: lib/ansible/modules/network/panos/panos_dag_tags.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        username=dict(default='admin'),
+        api_key=dict(no_log=True),
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_dag_tags.py]: No in docs

File: lib/ansible/modules/network/panos/panos_dag_tags.py (hunk 1 above)
Comment (gundalow) [reply lib/ansible/modules/network/panos/panos_dag_tags.py]: If you use an api_key is password still required?

You may wish to use `required_one_of=[['password', 'api_key']],` see https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec for more details

File: lib/ansible/modules/network/panos/panos_dag_tags.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    device = base.PanDevice.create_from_device(ip_address, username, password, api_key=api_key)
+
+    # If Panorama, validate the devicegroup
+    dev_group = None
+    if devicegroup and isinstance(device, panorama.Panorama):
+        dev_group = get_devicegroup(device, devicegroup)
+        if dev_group:
+            device.add(dev_group)
+        else:
+            module.fail_json(msg='\'%s\' device group not found in Panorama. Is the name correct?' % devicegroup)
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_dag_tags.py]: `msg="'%s' device group not found in Panorama. Is the name correct?" % devicegroup)`

File: lib/ansible/modules/network/panos/panos_dag_tags.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'metadata_version': '1.1'}
+
+DOCUMENTATION = '''
+---
+module: panos_dag_tags
+short_description: Create tags for DAG's
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_dag_tags.py]: `Create tags for DAG's on PAN-OS devices`

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/panos/panos_dag_tags.py (hunk 4 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_dag_tags.py]: ok

File: lib/ansible/modules/network/panos/panos_dag_tags.py (hunk 1 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_dag_tags.py]: OK, we used required_one before and this was oversight. Good catch and thank you.

File: lib/ansible/modules/network/panos/panos_dag_tags.py (hunk 3 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_dag_tags.py]: ok

File: lib/ansible/modules/network/panos/panos_dag_tags.py (hunk 2 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_dag_tags.py]: ok

**Comment by ivanbojer** [comment]
I've addressed all the changes. Thank you for the review.

**Comment by ivanbojer** [comment]
ready_for_review

**Comment by ivanbojer** [comment]
ready_for_review

**Comment by ivanbojer** [comment]
Didn't make it for 2.4

**Comment by gundalow** [comment]
@ivanbojer We can help you get this into Ansible 2.5 if you want to reopen the PR...

**Pull request closed without merging**